src/modules/groupMethods.ts

```diff
@@ -188,6 +190,76 @@ export const groupMethods = {
         return addCalculatedGroupFields(group, sessionUser);
     },
 
+    getGroupTree: async (data: GetGroupTree) => {
+        const { id, filters } = data;
+
+        const children = (id: string, supervisors?: string[], filterIds?: string[], defaultLevel = 1) =>
+            db
+                .withRecursive('tree', (qc) =>
+                    qc
+                        .selectFrom('Group')
+                        .select(({ fn }) => [
+                            'id',
+                            'parentId',
+                            'name',
+                            sql<number>`${defaultLevel}::int`.as('level'),
+                            fn.agg<string[]>('array_agg', ['parentId']).as('chain'),
+                        ])
+                        .where('id', '=', id)
+                        .where('archived', '=', false)
+                        .groupBy('id')
+                        .union((qb) =>
+                            qb
+                                .selectFrom('Group as g')
+                                .innerJoin('tree', 'tree.id', 'g.parentId')
+                                .select(({ fn, ref }) => [
+                                    ref('g.id').as('id'),
+                                    ref('g.parentId').as('parentId'),
+                                    ref('g.name').as('name'),
+                                    sql<number>`tree."level"::int + 1`.as('level'),
+                                    fn.agg<string[]>('array_append', ['tree.chain', 'g.parentId']).as('chain'),
+                                ])
+                                .where('archived', '=', false),
+                        ),
+                )
+                .selectFrom('Group as gr')
+                .innerJoin('User as u', (join) => join.onRef('u.id', '=', 'gr.supervisorId'))
+                .innerJoin('tree as tr', 'tr.id', 'gr.id')
+                .select(['gr.id', 'gr.name', 'tr.level', 'tr.chain', 'u.name as supervisorName'])
+                // @ts-ignore — supervisors checks in $if conditions
+                .$if(!!supervisors?.length, (qb) => qb.where('gr.supervisorId', 'in', supervisors))
+                // @ts-ignore — filterIds checks in $if conditions
+                .$if(!!filterIds?.length, (qb) => qb.where('gr.id', 'not in', filterIds))
+                .orderBy('tr.level')
+                .execute();
+
+        const filteredChildren = (await children(id, filters?.supervisors)).map((c) => ({ ...c, visible: true }));
+
+        if (!filters?.supervisors) return filteredChildren;
+
+        const filterIds: string[] = [];
+
+        filterIds.push(...filteredChildren.map(({ id }) => id));
+
+        const subchildren: {
+            id: string;
+            name: string;
+            level: number;
+            supervisorName: string | null;
+            chain: string[];
+            visible?: boolean;
+        }[] = [];
+
+        for (const child of filteredChildren) {
```

Comment:
Use `Promise.all` function instead of asynchronous `for..of` loop. Its will be a little bit faster, or you can make one more query which returns children project for each filtered projects